File: Lecture/3rd/Code/mlp.R
# x : n x L
# xb : n x (L+1) : add bias
# v : (L + 1) x M
# a : n x M
# ab : n x (M+1) : add bias
# w : (M + 1) x N
# wb : M x N : remove bias
# y : n x N
# t : n x N
# dh : n x M
# do : n x N

weights_init <- function(m, n) {
    w <- matrix(runif(m, -1, 1), m, 1)
    if(n==1) {
        return(w) # m x 1 matrix
    } else {
        for (i in 2:n) {
            w <- cbind(w, runif(m, -1, 1))
        }
        return(w) # m x n matrix
    }
}

sigmoid <- function(x) {
    return(1/(1 + exp(-x)))
}

forward <- function(weights, inputb) {
    s <- inputb %*% weights
    g <- sigmoid
    y <- g(s) # Numeric functions are automatically vectorized
    return(y)
}

addBias <- function(input, bias) {
    b <- matrix(bias, nrow(input), 1)
    return(cbind(b, input))
}

hideBias <- function(weight) {
    w <- weight[2:nrow(weight),]
    return(w)
}

train <- function(weights1, weights2, input, answer, eta = 0.25, times) {
    x <- input # x = n x L
    v <- weights1 # v = (L+1) x M
    w <- weights2 # w = (M+1) x N
    t <- answer # t = n x N
    xb <- addBias(x, -1) # xb = n x (L+1)
    for (i in 1:times) {
        a <- forward(v, xb) # a = n x M
        ab <- addBias(a, -1) # ab = n x (M+1)
        y <- forward(w, ab) # y = n x N
        err <- t(y - t) %*% (y - t) # err = n x n
        # print(err)
        wb <- hideBias(w) # Remove bias : wb = M x N
        do <- (y - t) * y * (1 - y) # do = n x N
        dh <- (do %*% t(wb)) * a * (1 - a) # dh = n x M
        w <- w - eta * (t(ab) %*% do)
        v <- v - eta * (t(xb) %*% dh)
    }
    # Recall
    a <- forward(v, xb)
    ab <- addBias(a,-1)
    y <- forward(w, ab)
    return(y)
}

# ==============================================================================
# Example - XOR
# ==============================================================================
v <- weights_init(3,2) # 3 x 2
w <- weights_init(3,1) # 3 x 1

x <- matrix(0, 4, 2) # 4 x 2
x[,1] = c(0, 0, 1, 1)
x[,2] = c(0, 1, 0, 1)

t <- c(0,1,1,0)

y <- train(v, w, x, t, 0.25, 10000)

print(y)

File: Lecture/3rd/Code/hw_julia/mlp.jl
using BenchmarkTools, LoopVectorization
#using Zygote

function weights_init(m, n)
    w = rand(m,n)
    return 2.0 * w .- 1.0
end

function sigmoid(x)
    return 1.0 ./ (1.0 .+ exp.(-x))
end

function sigmoid_s(x::T) where T <: Number
    1.0 / (1.0 + exp(-x))
end

#dsigmoid(x) = sigmoid'(x)

function sigmoid_avx(x::Matrix{Float64})
    (row, col) = size(x)
    m = Matrix{eltype(x)}(undef, row, col)
    @avx for i ∈ 1:row, j ∈ 1:col
        m[i, j] = sigmoid_s(x[i,j])
    end
    return m
end

function dsigmoid(x::Matrix{Float64})
    (row, col) = size(x)
    m = Matrix{eltype(x)}(undef, row, col)
    @avx for i in 1:row, j in 1:col
        m[i,j] = sigmoid_s(x[i,j]) * (one(eltype(x)) - sigmoid_s(x[i,j]))
    end
    return m
end

function forward(weight, inputb)
    s = inputb * weight
    return sigmoid_avx(s)
end

function add_bias(input, bias)
    b = repeat([bias], size(input, 1))
    return hcat(b, input)
end

function hide_bias(weight)
    return weight[2:end, :]
end

function train(w1, w2, input, answer, eta=0.25, times=20000)
    x = input
    v = w1
    w = w2
    t = answer
    xb = add_bias(x, -1.0)
    @inbounds for i in 2:times
        a = forward(v, xb)
        ab = add_bias(a, -1.0)
        y = forward(w, ab)
        wb = hide_bias(w)
        yt = y - t;
        dy = dsigmoid(y);
        d_o = yt .* dy;
        d_h = (d_o * wb') .* dsigmoid(a)
        
        w = w - eta * (ab' * d_o)
        v = v - eta * (xb' * d_h)
    end
    a = forward(v, xb)
    ab = add_bias(a, -1.0)
    y = forward(w, ab)
    return y
end

function main()
    v = weights_init(3,2)
    w = weights_init(3,1)
    
    x = Float64[0 0;0 1;1 0;1 1]
    t = Float64[0;1;1;0]
    
    #println("Input: ")
    #println(x)
    #println()
    #
    #println("True output: ")
    #println(t)
    #println()
    
    y = train(v, w, x, t)
    #println("Predict: ")
    #println(y)
    return y
end

#@benchmark main()

